# HW0916_02 CH02Q10

##### (a) To begin, load in the Boston data set, which is part of the ISLP library.

In [1]:
from ISLP import load_data
Boston = load_data('Boston')

##### (b) How many rows are in this data set? How many columns? What do the rows and columns represent?

In [2]:
Boston

crim    zn  indus  chas    nox     rm   age     dis  rad  tax  \
0    0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1  296   
1    0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2  242   
2    0.02729   0.0   7.07     0  0.469  7.185  61.1  4.9671    2  242   
3    0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3  222   
4    0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3  222   
..       ...   ...    ...   ...    ...    ...   ...     ...  ...  ...   
501  0.06263   0.0  11.93     0  0.573  6.593  69.1  2.4786    1  273   
502  0.04527   0.0  11.93     0  0.573  6.120  76.7  2.2875    1  273   
503  0.06076   0.0  11.93     0  0.573  6.976  91.0  2.1675    1  273   
504  0.10959   0.0  11.93     0  0.573  6.794  89.3  2.3889    1  273   
505  0.04741   0.0  11.93     0  0.573  6.030  80.8  2.5050    1  273   

     ptratio  lstat  medv  
0       15.3   4.98  24.0  
1       17.8   9.14  21.6  
2       17.8   4.03  34.7  
3       18.7   2.94  33.4  
4       18.7   5.33  36.2  
..       ...    ...   ...  
501     21.0   9.67  22.4  
502     21.0   9.08  20.6  
503     21.0   5.64  23.9  
504     21.0   6.48  22.0  
505     21.0   7.88  11.9  

[506 rows x 13 columns]

### 1.506rows
### 2.13columns
### 3.row:
##### crim: per capita crime rate by town.
##### zn: proportion of residential land zoned for lots over 25,000 sq.ft.
##### indus: proportion of non-retail business acres per town.
##### chas: Charles River dummy variable (= 1 if tract bounds river; 0 otherwise).
##### nox: nitrogen oxides concentration (parts per 10 million).
##### rm: average number of rooms per dwelling.
##### age: proportion of owner-occupied units built prior to 1940.
##### dis: weighted mean of distances to five Boston employment centres.
##### rad: index of accessibility to radial highways.
##### tax: full-value property-tax rate per 10,000 dollars.
##### ptratio: pupil-teacher ratio by town.
##### lstat: lower status of the population (percent).
##### medv: median value of owner-occupied homes in 1000 dollars.
### column:
##### suburbs of Boston

##### (c) Make some pairwise scatterplots of the predictors (columns) in this data set. Describe your findings.

In [3]:
import seaborn as sns
import matplotlib.pyplot as plt
import pandas as pd
sns.pairplot(Boston)

![Alt text](pair.png)

In [4]:
Boston.corr()["medv"]

crim      -0.388305
zn         0.360445
indus     -0.483725
chas       0.175260
nox       -0.427321
rm         0.695360
age       -0.376955
dis        0.249929
rad       -0.381626
tax       -0.468536
ptratio   -0.507787
lstat     -0.737663
medv       1.000000
Name: medv, dtype: float64

##### rm and medv are highly positively correlated
##### lstat and medv are highly negatively correlated

##### (d) Are any of the predictors associated with per capita crime rate? If so, explain the relationship.

In [5]:
Boston.describe()
sns.pairplot(Boston2)

crim          zn       indus        chas         nox          rm  \
count  506.000000  506.000000  506.000000  506.000000  506.000000  506.000000   
mean     3.613524   11.363636   11.136779    0.069170    0.554695    6.284634   
std      8.601545   23.322453    6.860353    0.253994    0.115878    0.702617   
min      0.006320    0.000000    0.460000    0.000000    0.385000    3.561000   
25%      0.082045    0.000000    5.190000    0.000000    0.449000    5.885500   
50%      0.256510    0.000000    9.690000    0.000000    0.538000    6.208500   
75%      3.677083   12.500000   18.100000    0.000000    0.624000    6.623500   
max     88.976200  100.000000   27.740000    1.000000    0.871000    8.780000   

              age         dis         rad         tax     ptratio       lstat  \
count  506.000000  506.000000  506.000000  506.000000  506.000000  506.000000   
mean    68.574901    3.795043    9.549407  408.237154   18.455534   12.653063   
std     28.148861    2.105710    8.707259  168.537116    2.164946    7.141062   
min      2.900000    1.129600    1.000000  187.000000   12.600000    1.730000   
25%     45.025000    2.100175    4.000000  279.000000   17.400000    6.950000   
50%     77.500000    3.207450    5.000000  330.000000   19.050000   11.360000   
75%     94.075000    5.188425   24.000000  666.000000   20.200000   16.955000   
max    100.000000   12.126500   24.000000  711.000000   22.000000   37.970000   

             medv  
count  506.000000  
mean    22.532806  
std      9.197104  
min      5.000000  
25%     17.025000  
50%     21.200000  
75%     25.000000  
max     50.000000

![Alt text](Boston2.png)

In [6]:
Boston2 = Boston[Boston["crim"] > 1]
Boston2.describe()

crim     zn       indus        chas         nox          rm  \
count  174.000000  174.0  174.000000  174.000000  174.000000  174.000000   
mean    10.138975    0.0   17.836437    0.086207    0.677006    6.021885   
std     12.272585    0.0    2.454685    0.281479    0.088737    0.732935   
min      1.002450    0.0    8.140000    0.000000    0.532000    3.561000   
25%      3.489463    0.0   18.100000    0.000000    0.605000    5.656750   
50%      6.759445    0.0   18.100000    0.000000    0.679000    6.113000   
75%     12.023925    0.0   18.100000    0.000000    0.713000    6.404750   
max     88.976200    0.0   21.890000    1.000000    0.871000    8.780000   

              age         dis         rad         tax     ptratio       lstat  \
count  174.000000  174.000000  174.000000  174.000000  174.000000  174.000000   
mean    90.339655    2.124741   19.344828  597.373563   19.294253   17.815115   
std     13.010515    0.726156    8.279549  123.799594    2.117576    7.399529   
min     29.300000    1.129600    4.000000  304.000000   14.700000    1.730000   
25%     87.900000    1.590025   24.000000  666.000000   20.200000   13.137500   
50%     94.850000    1.951200   24.000000  666.000000   20.200000   17.185000   
75%     98.775000    2.428825   24.000000  666.000000   20.200000   22.705000   
max    100.000000    4.498600   24.000000  666.000000   21.200000   37.970000   

             medv  
count  174.000000  
mean    17.613793  
std      9.361360  
min      5.000000  
25%     12.525000  
50%     15.200000  
75%     20.600000  
max     50.000000

##### In high per capita crime rate(1%) suburbs of Boston, there are some features of surbubs
1.Most of them rad are high(24)(Max of full sample:24)\
2.Most of them tax are high(666)(Max of full sample:711)

3.indus are higher(18.1)(75th percentile of full sample)\
4.ptratio are higher(20.2)(75th percentile of full sample)
##### It seems that 3. and 4. not really a significant features, but just a statistical results.

##### (e) Do any of the suburbs of Boston appear to have particularly high crime rates? Tax rates? Pupil-teacher ratios? Comment on the range of each predictor.

In [7]:
Boston.describe()

crim          zn       indus        chas         nox          rm  \
count  506.000000  506.000000  506.000000  506.000000  506.000000  506.000000   
mean     3.613524   11.363636   11.136779    0.069170    0.554695    6.284634   
std      8.601545   23.322453    6.860353    0.253994    0.115878    0.702617   
min      0.006320    0.000000    0.460000    0.000000    0.385000    3.561000   
25%      0.082045    0.000000    5.190000    0.000000    0.449000    5.885500   
50%      0.256510    0.000000    9.690000    0.000000    0.538000    6.208500   
75%      3.677083   12.500000   18.100000    0.000000    0.624000    6.623500   
max     88.976200  100.000000   27.740000    1.000000    0.871000    8.780000   

              age         dis         rad         tax     ptratio       lstat  \
count  506.000000  506.000000  506.000000  506.000000  506.000000  506.000000   
mean    68.574901    3.795043    9.549407  408.237154   18.455534   12.653063   
std     28.148861    2.105710    8.707259  168.537116    2.164946    7.141062   
min      2.900000    1.129600    1.000000  187.000000   12.600000    1.730000   
25%     45.025000    2.100175    4.000000  279.000000   17.400000    6.950000   
50%     77.500000    3.207450    5.000000  330.000000   19.050000   11.360000   
75%     94.075000    5.188425   24.000000  666.000000   20.200000   16.955000   
max    100.000000   12.126500   24.000000  711.000000   22.000000   37.970000   

             medv  
count  506.000000  
mean    22.532806  
std      9.197104  
min      5.000000  
25%     17.025000  
50%     21.200000  
75%     25.000000  
max     50.000000

In [8]:
Boston.sort_values("crim", ascending=False)["crim"][:len(Boston)//20].describe()

count    25.000000
mean     32.333888
std      19.475929
min      15.874400
25%      19.609100
50%      24.393800
75%      38.351800
max      88.976200
Name: crim, dtype: float64

### high crime rates
##### Yes. We calculate mean of  highest 5% crime rates is 32.333888, it's different from 10.138975 of full sample.
##### Max of highest 5% crime rates is 88.976200, min of them is 15.874400.

In [9]:
Boston.sort_values("tax", ascending=False)["tax"][:len(Boston)//20].describe()

count     25.000000
mean     675.000000
std       18.371173
min      666.000000
25%      666.000000
50%      666.000000
75%      666.000000
max      711.000000
Name: tax, dtype: float64

### high Tax rates
##### Yes. We calculate mean of  highest 5% crime rates is 675, it's different from 408.237154 of full sample.
##### Max of highest 5% tax rates is 711, min of them is 666.

In [10]:
Boston.sort_values("ptratio", ascending=False)["ptratio"][:len(Boston)//20].describe()

count    25.000000
mean     21.204000
std       0.255734
min      21.000000
25%      21.000000
50%      21.200000
75%      21.200000
max      22.000000
Name: ptratio, dtype: float64

### high Pupil-teacher ratios
##### No. We calculate mean of  highest 5% crime rates is 21.204000, it's not much different from 18.455534 of full sample.
##### Max of highest 5% pupil-teacher ratios is 22, min of them is 21.

##### (f) How many of the suburbs in this data set bound the Charles river?

In [11]:
sum(Boston["chas"]==1)

35

##### (g) What is the median pupil-teacher ratio among the towns in this data set?

In [12]:
Boston["ptratio"].median()

19.05

##### (h) Which suburb of Boston has lowest median value of owner occupied homes? What are the values of the other predictors for that suburb, and how do those values compare to the overall ranges for those predictors? Comment on your findings.

In [13]:
Boston.sort_values("medv").head()

crim   zn  indus  chas    nox     rm    age     dis  rad  tax  \
398  38.35180  0.0   18.1     0  0.693  5.453  100.0  1.4896   24  666   
405  67.92080  0.0   18.1     0  0.693  5.683  100.0  1.4254   24  666   
400  25.04610  0.0   18.1     0  0.693  5.987  100.0  1.5888   24  666   
399   9.91655  0.0   18.1     0  0.693  5.852   77.8  1.5004   24  666   
414  45.74610  0.0   18.1     0  0.693  4.519  100.0  1.6582   24  666   

     ptratio  lstat  medv  
398     20.2  30.59   5.0  
405     20.2  22.98   5.0  
400     20.2  26.77   5.6  
399     20.2  29.97   6.3  
414     20.2  36.98   7.0

##### 398 and 405 suburbs have lowest median value of owner occupied homes
##### 5.0

In [14]:
Boston[Boston["medv"]==5]

crim   zn  indus  chas    nox     rm    age     dis  rad  tax  \
398  38.3518  0.0   18.1     0  0.693  5.453  100.0  1.4896   24  666   
405  67.9208  0.0   18.1     0  0.693  5.683  100.0  1.4254   24  666   

     ptratio  lstat  medv  
398     20.2  30.59   5.0  
405     20.2  22.98   5.0

In [15]:
Boston.describe()

crim          zn       indus        chas         nox          rm  \
count  506.000000  506.000000  506.000000  506.000000  506.000000  506.000000   
mean     3.613524   11.363636   11.136779    0.069170    0.554695    6.284634   
std      8.601545   23.322453    6.860353    0.253994    0.115878    0.702617   
min      0.006320    0.000000    0.460000    0.000000    0.385000    3.561000   
25%      0.082045    0.000000    5.190000    0.000000    0.449000    5.885500   
50%      0.256510    0.000000    9.690000    0.000000    0.538000    6.208500   
75%      3.677083   12.500000   18.100000    0.000000    0.624000    6.623500   
max     88.976200  100.000000   27.740000    1.000000    0.871000    8.780000   

              age         dis         rad         tax     ptratio       lstat  \
count  506.000000  506.000000  506.000000  506.000000  506.000000  506.000000   
mean    68.574901    3.795043    9.549407  408.237154   18.455534   12.653063   
std     28.148861    2.105710    8.707259  168.537116    2.164946    7.141062   
min      2.900000    1.129600    1.000000  187.000000   12.600000    1.730000   
25%     45.025000    2.100175    4.000000  279.000000   17.400000    6.950000   
50%     77.500000    3.207450    5.000000  330.000000   19.050000   11.360000   
75%     94.075000    5.188425   24.000000  666.000000   20.200000   16.955000   
max    100.000000   12.126500   24.000000  711.000000   22.000000   37.970000   

             medv  
count  506.000000  
mean    22.532806  
std      9.197104  
min      5.000000  
25%     17.025000  
50%     21.200000  
75%     25.000000  
max     50.000000

Finding\
    crim:398 and 405 are over 75th percentile(3.677083) of overall predictors.\
    zn:398 and 405 are over 75th percentile(12.5) of overall predictors.\
    indus:398 and 405 are on 75th percentile(18.1) of overall predictors.\
    chas:398 and 405 are not bound the Charles river.\
    nox:398 and 405 are over 75th percentile(0.624) of overall predictors.\
    rm:398 and 405 are below the 25th percentile(5.8855) of overall predictors.\
    dis:398 and 405 are below the 25th percentile(2.100175) of overall predictors.\
    rad:398 and 405 are the max of overall predictors.\
    tax:398 and 405 are on 75th percentile(666) of overall predictors.\
    ptratio:398 and 405 are on 75th percentile(20.2) of overall predictors.\
    lstat:398 and 405 are over 75th percentile(16.955) of overall predictors.

##### (i) In this data set, how many of the suburbs average more than seven rooms per dwelling? More than eight rooms per dwelling? Comment on the suburbs that average more than eight rooms per dwelling.

In [16]:
sum(Boston["rm"] > 7)

64

In [17]:
sum(Boston["rm"] > 8)

13

In [18]:
Boston[Boston["rm"] > 8]

crim    zn  indus  chas     nox     rm   age     dis  rad  tax  \
97   0.12083   0.0   2.89     0  0.4450  8.069  76.0  3.4952    2  276   
163  1.51902   0.0  19.58     1  0.6050  8.375  93.9  2.1620    5  403   
204  0.02009  95.0   2.68     0  0.4161  8.034  31.9  5.1180    4  224   
224  0.31533   0.0   6.20     0  0.5040  8.266  78.3  2.8944    8  307   
225  0.52693   0.0   6.20     0  0.5040  8.725  83.0  2.8944    8  307   
226  0.38214   0.0   6.20     0  0.5040  8.040  86.5  3.2157    8  307   
232  0.57529   0.0   6.20     0  0.5070  8.337  73.3  3.8384    8  307   
233  0.33147   0.0   6.20     0  0.5070  8.247  70.4  3.6519    8  307   
253  0.36894  22.0   5.86     0  0.4310  8.259   8.4  8.9067    7  330   
257  0.61154  20.0   3.97     0  0.6470  8.704  86.9  1.8010    5  264   
262  0.52014  20.0   3.97     0  0.6470  8.398  91.5  2.2885    5  264   
267  0.57834  20.0   3.97     0  0.5750  8.297  67.0  2.4216    5  264   
364  3.47428   0.0  18.10     1  0.7180  8.780  82.9  1.9047   24  666   

     ptratio  lstat  medv  
97      18.0   4.21  38.7  
163     14.7   3.32  50.0  
204     14.7   2.88  50.0  
224     17.4   4.14  44.8  
225     17.4   4.63  50.0  
226     17.4   3.13  37.6  
232     17.4   2.47  41.7  
233     17.4   3.95  48.3  
253     19.1   3.54  42.8  
257     13.0   5.12  50.0  
262     13.0   5.91  48.8  
267     13.0   7.44  50.0  
364     20.2   5.29  21.9

In [19]:
Boston[Boston["rm"] > 8].describe()

crim         zn      indus       chas        nox         rm  \
count  13.000000  13.000000  13.000000  13.000000  13.000000  13.000000   
mean    0.718795  13.615385   7.078462   0.153846   0.539238   8.348538   
std     0.901640  26.298094   5.392767   0.375534   0.092352   0.251261   
min     0.020090   0.000000   2.680000   0.000000   0.416100   8.034000   
25%     0.331470   0.000000   3.970000   0.000000   0.504000   8.247000   
50%     0.520140   0.000000   6.200000   0.000000   0.507000   8.297000   
75%     0.578340  20.000000   6.200000   0.000000   0.605000   8.398000   
max     3.474280  95.000000  19.580000   1.000000   0.718000   8.780000   

             age        dis        rad         tax    ptratio      lstat  \
count  13.000000  13.000000  13.000000   13.000000  13.000000  13.000000   
mean   71.538462   3.430192   7.461538  325.076923  16.361538   4.310000   
std    24.608723   1.883955   5.332532  110.971063   2.410580   1.373566   
min     8.400000   1.801000   2.000000  224.000000  13.000000   2.470000   
25%    70.400000   2.288500   5.000000  264.000000  14.700000   3.320000   
50%    78.300000   2.894400   7.000000  307.000000  17.400000   4.140000   
75%    86.500000   3.651900   8.000000  307.000000  17.400000   5.120000   
max    93.900000   8.906700  24.000000  666.000000  20.200000   7.440000   

            medv  
count  13.000000  
mean   44.200000  
std     8.092383  
min    21.900000  
25%    41.700000  
50%    48.300000  
75%    50.000000  
max    50.000000

Comment\
    crim:These dwelling are lower crime rates than overall predictors.\
    zn:Not too much difference between these dwelling and overall predictors.\
    indus:Little lower than overall predictors.\
    chas:2 of them bound the Charles river.\
    nox:Not too much difference between these dwelling and overall predictors.\
    dis:Not too much difference between these dwelling and overall predictors.\
    rad:Little lower than overall predictors.\
    tax:Little lower than overall predictors.\
    ptratio:Not too much difference between these dwelling and overall predictors.\
    lstat:These dwelling are less lower status of the population than overall predictors.\
    medv:These dwelling are higher median value of owner-occupied homes in 1000 dollars.